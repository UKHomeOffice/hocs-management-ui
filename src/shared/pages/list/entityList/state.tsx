import EntityListItem from '../../../models/entityListItem';

export interface State {
    entities: EntityListItem[];
    entitiesLoaded: boolean;
}
