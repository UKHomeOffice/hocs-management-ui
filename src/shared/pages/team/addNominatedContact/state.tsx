import Item from '../../../models/item';
export interface State {
    inputValue: string;
    selectedContacts: Item[];
    teamName?: string;
    teamUUID: string;
}
