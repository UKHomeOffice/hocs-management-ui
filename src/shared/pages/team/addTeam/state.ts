import Team from '../../../models/team';

export interface State {
    team: Team;
}
