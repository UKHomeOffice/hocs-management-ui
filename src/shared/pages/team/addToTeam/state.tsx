import Item from '../../../models/item';
export interface State {
    inputValue: string;
    selectedUser?: Item | string;
    selectedUsers: Item[];
    teamName?: string;
    users: Item[];
}
