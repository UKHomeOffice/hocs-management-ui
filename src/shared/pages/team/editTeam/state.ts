import Item from '../../../models/item';

export interface State {
    currentDisplayName: string,
    newDisplayName: string
    unit?: Item;
    initialUnit?: Item;
}
