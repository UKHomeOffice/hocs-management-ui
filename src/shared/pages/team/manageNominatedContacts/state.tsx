import Item from '../../../models/item';
export interface State {
    contacts: Item[];
}
