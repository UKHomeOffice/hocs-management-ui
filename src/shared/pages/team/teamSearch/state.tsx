export interface State {
    teamUUID: string;
}
