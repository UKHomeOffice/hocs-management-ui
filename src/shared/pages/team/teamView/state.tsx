import Item from '../../../models/item';

export interface State {
    teamMembers: Item[];
    teamMembersLoaded: boolean;
    teamName?: string;
    unitName?: string;
    active?: boolean;
}
