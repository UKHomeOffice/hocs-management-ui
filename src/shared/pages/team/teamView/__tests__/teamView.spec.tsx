import React from 'react';
import { match } from 'react-router';
import { createBrowserHistory, History, Location } from 'history';
import { act, wait, fireEvent, getByText, render, RenderResult } from '@testing-library/react';
import TeamView from '../teamView';
import * as TeamsService from '../../../../services/teamsService';
import * as UsersService from '../../../../services/usersService';
import { State } from '../state';

let match: match<any>;
let history: History<any>;
let location: Location;
let mockState: State;

jest.mock('../../../../services/teamsService', () => ({
    __esModule: true,
    getTeamMembers: jest.fn().mockReturnValue(Promise.resolve({
        data: [{
            label: '__user1__',
            value: '__userId1__'
        }, {
            label: '__user2__',
            value: '__userId2__'
        }]
    })),
    getTeam: jest.fn().mockReturnValue(Promise.resolve({
        active: true,
        displayName: '__displayName__',
        letterName: '__letterName__',
        permissions: [],
        type: '__type__'
    }))
}));

jest.mock('../../../../services/usersService', () => ({
    __esModule: true,
    addUserToTeam: jest.fn().mockReturnValue(Promise.resolve()),
    deleteUserFromTeam: jest.fn().mockReturnValue(Promise.resolve()),
    getUsers: jest.fn().mockReturnValue(Promise.resolve({
        data: [{
            label: '__user1__',
            value: '__userId1__'
        }, {
            label: '__user2__',
            value: '__userId2__'
        }]
    }))
}));

const getTeamSpy = jest.spyOn(TeamsService, 'getTeam');
const getTeamMembersSpy = jest.spyOn(TeamsService, 'getTeamMembers');
const deleteUserFromTeamSpy = jest.spyOn(UsersService, 'deleteUserFromTeam');
const useReducerSpy = jest.spyOn(React, 'useReducer');
const dispatch = jest.fn();

beforeEach(() => {
    history = createBrowserHistory();
    match = {
        isExact: true,
        params: { teamId: '__teamId__' },
        path: '',
        url: ''
    };

    location = {
        hash: '',
        key: '',
        pathname: '',
        search: '',
        state: {}
    };
    mockState = {
        errorDescription: '',
        errorTitle: '',
        teamMembersLoaded: true,
        teamMembers: [{
            label: '__user1__',
            value: '__userId1__'
        }, {
            label: '__user2__',
            value: '__userId2__'
        }],
        teamName: '__teamName__'
    };
    useReducerSpy.mockImplementationOnce(() => [mockState, dispatch]);
    dispatch.mockReset();
});

describe('when the teamView component is mounted', () => {
    it('should render with default props', async () => {
        expect.assertions(3);
        let wrapper: RenderResult;
        act(() => {
            wrapper = render(<TeamView history={history} location={location} match={match}></TeamView>);
        });

        await wait(() => {
            expect(getTeamSpy).toHaveBeenCalled();
            expect(getTeamMembersSpy).toHaveBeenCalled();
            expect(wrapper.container).toMatchSnapshot();
        });
    });
});

describe('when the back button is clicked', () => {
    it('should push a new page into the history', async () => {
        history.push = jest.fn();
        let wrapper: RenderResult;
        act(() => {
            wrapper = render(<TeamView history={history} location={location} match={match}></TeamView>);
        });

        await wait(async () => {
            const backButton = getByText(wrapper.container, 'Back');
            fireEvent.click(backButton);
        });

        expect(history.push).toHaveBeenCalledWith('/team-search');
    });
});

describe('when the Add team members button is clicked', () => {
    it('should push a new page into the history', async () => {
        history.push = jest.fn();
        let wrapper: RenderResult;
        act(() => {
            wrapper = render(<TeamView history={history} location={location} match={match}></TeamView>);
        });

        await wait(async () => {
            const addTeamMembersButton = getByText(wrapper.container, 'Add team members');
            fireEvent.click(addTeamMembersButton);
        });

        expect(history.push).toHaveBeenCalledWith('/team/__teamId__/add-users');
    });
});

describe('when the remove user button is clicked', () => {
    it('should remove the row from the users table', async () => {
        let wrapper: RenderResult;
        act(() => {
            wrapper = render(<TeamView history={history} location={location} match={match}></TeamView>);
        });

        await wait(async () => {
            const selectedUser = getByText(wrapper.container, '__user1__');
            dispatch.mockReset();
            const row = (selectedUser.closest('tr'));
            const removeButton = getByText(row as HTMLElement, 'Remove');
            fireEvent.click(removeButton);
        });

        expect(deleteUserFromTeamSpy).nthCalledWith(1,  '__userId1__', '__teamId__');
        expect(getTeamMembersSpy).nthCalledWith(1, '__teamId__');

    });
});
