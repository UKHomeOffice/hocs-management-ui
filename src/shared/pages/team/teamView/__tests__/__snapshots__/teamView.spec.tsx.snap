// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when the teamView component is mounted with RENAME_TEAM role should render with default props 1`] = `
<div>
  <div
    class="govuk-form-group"
  >
    <a
      class="govuk-back-link"
      href="/team-search"
    >
      Back
    </a>
    <div>
      <h1
        class="govuk-heading-xl"
      >
        View and remove team members
      </h1>
      <h2
        class="govuk-heading-l"
      >
        Team: __teamName__
      </h2>
      <div>
        <table
          class="govuk-table"
        >
          <thead
            class="govuk-table__head"
          >
            <tr
              class="govuk-table__row"
            >
              <th
                class="govuk-table__header"
                scope="col"
              >
                Team members
              </th>
              <th
                class="govuk-table__header"
                scope="col"
              >
                Action
              </th>
            </tr>
          </thead>
          <tbody
            class="govuk-table__body"
          >
            <tr
              class="govuk-table__row"
            >
              <td
                class="govuk-table__cell"
              >
                __user1__
              </td>
              <td
                class="govuk-table__cell"
              >
                <a
                  href="#"
                >
                  Remove
                </a>
              </td>
            </tr>
            <tr
              class="govuk-table__row"
            >
              <td
                class="govuk-table__cell"
              >
                __user2__
              </td>
              <td
                class="govuk-table__cell"
              >
                <a
                  href="#"
                >
                  Remove
                </a>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <button
        class="govuk-button govuk-!-margin-right-1 add-team-members-button"
        data-module="govuk-button"
        type="submit"
      >
        Add team members
      </button>
      <button
        class="govuk-button govuk-!-margin-right-1 govuk-button--secondary manage-nominated-contacts-button"
        data-module="govuk-button"
        type="submit"
      >
        Manage nominated contacts
      </button>
      <button
        class="govuk-button govuk-button--secondary"
        data-module="govuk-button"
        type="submit"
      >
        Edit Team
      </button>
    </div>
  </div>
</div>
`;

exports[`when the teamView component is mounted without RENAME_TEAM role should render with default props 1`] = `
<div>
  <div
    class="govuk-form-group"
  >
    <a
      class="govuk-back-link"
      href="/team-search"
    >
      Back
    </a>
    <div>
      <h1
        class="govuk-heading-xl"
      >
        View and remove team members
      </h1>
      <h2
        class="govuk-heading-l"
      >
        Team: __teamName__
      </h2>
      <div>
        <table
          class="govuk-table"
        >
          <thead
            class="govuk-table__head"
          >
            <tr
              class="govuk-table__row"
            >
              <th
                class="govuk-table__header"
                scope="col"
              >
                Team members
              </th>
              <th
                class="govuk-table__header"
                scope="col"
              >
                Action
              </th>
            </tr>
          </thead>
          <tbody
            class="govuk-table__body"
          >
            <tr
              class="govuk-table__row"
            >
              <td
                class="govuk-table__cell"
              >
                __user1__
              </td>
              <td
                class="govuk-table__cell"
              >
                <a
                  href="#"
                >
                  Remove
                </a>
              </td>
            </tr>
            <tr
              class="govuk-table__row"
            >
              <td
                class="govuk-table__cell"
              >
                __user2__
              </td>
              <td
                class="govuk-table__cell"
              >
                <a
                  href="#"
                >
                  Remove
                </a>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <button
        class="govuk-button govuk-!-margin-right-1 add-team-members-button"
        data-module="govuk-button"
        type="submit"
      >
        Add team members
      </button>
      <button
        class="govuk-button govuk-!-margin-right-1 govuk-button--secondary manage-nominated-contacts-button"
        data-module="govuk-button"
        type="submit"
      >
        Manage nominated contacts
      </button>
    </div>
  </div>
</div>
`;
